File: delete a dll.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* prev;
    struct Node* next;
};

struct Node* deleteNode(struct Node* head, int key) {
    struct Node* temp = head;
    while (temp != NULL && temp->data != key)
        temp = temp->next;

    if (temp == NULL)
        return head; // Not found

    if (temp->prev == NULL) {
        head = temp->next;
        if (head != NULL)
            head->prev = NULL;
    } else {
        temp->prev->next = temp->next;
    }

    if (temp->next != NULL)
        temp->next->prev = temp->prev;

    free(temp);
    return head;
}

void printDLL(struct Node* head) {
    while (head != NULL) {
        printf("%d <-> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->prev = NULL;
    node->next = NULL;
    return node;
}

int main() {
    struct Node* head = newNode(10);
    struct Node* second = newNode(20);
    struct Node* third = newNode(30);

    head->next = second;
    second->prev = head;
    second->next = third;
    third->prev = second;

    printf("Original DLL:\n");
    printDLL(head);

    head = deleteNode(head, 20);

    printf("After deleting 20:\n");
    printDLL(head);

    return 0;
}

File: delete a dll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def delete_node(head, key):
    temp = head

    while temp and temp.data != key:
        temp = temp.next

    if not temp:
        return head  # Key not found
    if temp.prev is None:
        head = temp.next
        if head:
            head.prev = None
    else:
        temp.prev.next = temp.next

    if temp.next:
        temp.next.prev = temp.prev

    return head

def print_dll(head):
    while head:
        print(head.data, end=" <-> ")
        head = head.next
    print("NULL")


print("Original DLL:")
print_dll(head)

head = delete_node(head, 20)
print("After deleting 20:")
print_dll(head)
